### PR TITLE
Editor integration tests

### PR DESCRIPTION
**Pull request by apedroferreira**
Closes https://github.com/mui/mui-toolpad/issues/758

Refactored to use Playright's Page Object Model as we intended.

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/material-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

### CODE REVIEW

File: packages/toolpad-app/src/runtime/ToolpadApp.tsx [hunk 1]
```diff
@@ -307,7 +307,7 @@ function PageRoot({ children }: PageRootProps) {
   return (
     <Container>
       <Stack
-        data-testid="page-root"
+        data-test-id="page-root"
```
Comment (apedroferreira) [inline packages/toolpad-app/src/runtime/ToolpadApp.tsx:310]: this one really tripped me up :D

Comment (Janpot) [reply packages/toolpad-app/src/runtime/ToolpadApp.tsx]: The reason it's called that way is so we can also use it for `getByTestId`: https://testing-library.com/docs/queries/bytestid/#api

Comment (apedroferreira) [reply packages/toolpad-app/src/runtime/ToolpadApp.tsx]: yeah, there another in another file that was `data-test-id` so i was trying that first, i guess we can use `data-testid` everywhere

**Comment by apedroferreira** [comment]
Looks like we have to use the overlay element to capture clicks/drags, as it's the one that captures them. It doesn't work with other elements in the same position.
Also I don't plan to add lots of test ids but I think that for these specific elements it was the best choice.

Let me know if you think anything should be different in terms of how i reorganized the files and their structure a bit too!

**Comment by apedroferreira** [comment]
Will add a couple of tests for prop controls in a separate PR now too.

File: test/integration/editor.spec.ts [hunk 2]
```diff
@@ -0,0 +1,154 @@
+import { test, expect, Page, Locator } from '@playwright/test';
+import createApp from '../utils/createApp';
+import generateId from '../utils/generateId';
+import { canvasFrame, componentCatalog, pageRoot, pageOverlay } from '../utils/locators';
+import twoFieldsDomTemplate from '../domTemplates/twoTextFields.json';
```
Comment (Janpot) [inline test/integration/editor.spec.ts:5]: I prefer not to organize this using a big `/domTemplates/...` folder containing doms for all the tests, that doesn't scale well. The json should just live right next to the test. Can we rather do something like the following?

```
test
  integration
    appRename
      index.spec.ts
    editor
      index.spec.ts
      input.json
    smoke
      index.spec.ts
```

File: test/integration/editor.spec.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+  const canvasPageOverlayBoundingBox = await canvasPageOverlayLocator.boundingBox();
+
+  const secondTextFieldLocator = canvasInputLocator.nth(1);
+  const secondTextFieldBoundingBox = await secondTextFieldLocator.boundingBox();
+
+  expect(canvasPageOverlayBoundingBox).toBeDefined();
+  expect(secondTextFieldBoundingBox).toBeDefined();
+
+  await canvasMoveElementHandleLocator.dragTo(canvasPageOverlayLocator, {
```
Comment (Janpot) [inline test/integration/editor.spec.ts:101]: Can we use mouse events to simulate drag and drop instead? Did you try the strategy from https://github.com/mui/mui-toolpad/issues/758#issuecomment-1207914180 ?

File: test/integration/editor.spec.ts (hunk 2 above)
Comment (apedroferreira) [reply test/integration/editor.spec.ts:5]: sure, that sounds fine too

File: test/integration/editor.spec.ts (hunk 3 above)
Comment (apedroferreira) [reply test/integration/editor.spec.ts:101]: Probably can do done like those, I just thought `dragTo` would be better since it encapsulates the whole drag interaction?
Also I can add a comment here explaining that it's dragging to the right side of the other element.

Comment (Janpot) [reply test/integration/editor.spec.ts:101]: It seems to me that the method as used in the core is much closer to real user interaction, and less dependant on implementation details

Comment (apedroferreira) [reply test/integration/editor.spec.ts:101]: ok, ill also add a method about getting an (x,y) position relative to the overlay, it should make this kind of code more readable.

Comment (Janpot) [reply test/integration/editor.spec.ts:101]: Best would be to avoid the overlay whenever we can and rather detect/calculate based these things on the underlying app running in the iframe. The overlay is an implementation detail, we may want to change its implementation in the future. In fact I'd say it's very likely we will heavily change this implementation at some point.

Comment (apedroferreira) [reply test/integration/editor.spec.ts:101]: Yeah, I agree, the problem is that the overlay is handling all the events, I tried with other elements in the same position but it did not work. I'd be glad to do it in an alternative way but would have to look more into what else is possible.

Comment (apedroferreira) [reply test/integration/editor.spec.ts]: alright i hadn't realized we could use the Playwright's page object and do everything relative to it - that means i don't have to use the overlay after all!
thanks for the feedback, it was super helpful, this should be better now, lmk if you see anything else that can be improved

File: .circleci/config.yml [hunk 4]
```diff
@@ -170,7 +170,7 @@ jobs:
       - run:
           command: |
             # See https://circleci.com/docs/2.0/env-vars/#setting-an-environment-variable-in-a-shell-command
-            TAG=$CIRCLE_SHA1 docker-compose -f ./docker/compose/docker-compose.yml up -d
+            TAG=$CIRCLE_SHA1 docker-compose -f ./docker/compose/docker-compose.test.yml up -d
```
Comment (Janpot) [inline .circleci/config.yml:173]: perhaps we should move this top-level, next to docker-compose.test.yml. We could reserve `./docker/compose/` for public examples?

File: test/integration/editor/index.spec.ts [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+  await page.mouse.up();
+}
+
+test('can place new components from catalog', async ({ page }) => {
+  const appId = generateId();
+
+  await page.goto('/');
+  await createApp(page, `App ${appId}`);
+
+  const canvasFrameLocator = page.frameLocator(canvasFrame);
```
Comment (Janpot) [inline test/integration/editor/index.spec.ts:35]: For the future we could think of an [Object model](https://playwright.dev/docs/test-pom) for this. Been trying that out in https://github.com/mui/mui-toolpad/pull/838
Doesn't have to be change for this PR

File: test/utils/createApp.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
+import { Page } from '@playwright/test';
+
+export default async function createApp(page: Page, name: string, dom?: string) {
+  await page.locator('button:has-text("create new")').click();
+
+  await page.fill('[role="dialog"] label:has-text("name")', name);
```
Comment (Janpot) [inline test/utils/createApp.ts:6]: We could improve dialog selectors a bit: https://github.com/mui/mui-toolpad/pull/838/files#diff-e537b99227068da9665d973e6573a1cd6da8d36be778d2aba43b159890a43d99R30

But we can merge this as is and use the ToolpadHome model as an optimization later

**Review by Janpot — APPROVED** [review]
I'd move the compose file top-level, but other than that we can merge this 👍

**Comment by apedroferreira** [comment]
> I'd move the compose file top-level, but other than that we can merge this 👍
> 
> _edit_ 🤔 wait, the test doesn't pass?

yeah, the CI with the env var isn't working so something must be wrong there, and there's a separate issue with the Firefox tests that include drag & drop not passing

File: test/integration/editor/index.spec.ts (hunk 5 above)
Comment (apedroferreira) [reply test/integration/editor/index.spec.ts:35]: didn't know about it, yeah we could use to abstract the things we end up reusing a lot
i might still include it in this PR when we take a look at the firefox stuff again, shouldn't be too time consuming

File: test/utils/createApp.ts (hunk 6 above)
Comment (apedroferreira) [reply test/utils/createApp.ts:6]: sure, i just took this function from the other tests where it was
we can use that model if we merge it first

File: docker/compose/docker-compose.test.yml [hunk 7]
```diff
@@ -0,0 +1,30 @@
+version: '3.7'
```
Comment (Janpot) [inline docker/compose/docker-compose.test.yml:1]: This one is not needed anymore I think

File: test/integration/propControls/index.spec.ts [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    .elementHandle();
+  const propControlLabelFor = await propControlLabelHandle?.getAttribute('for');
+
+  return selectedNodeEditorLocator.locator(`input[id="${propControlLabelFor}"]`);
+}
+
+async function getInputElementLabelLocator(page: Page, inputLocator: Locator) {
+  const canvasFrameLocator = page.frameLocator(canvasFrame);
+
+  const inputHandle = await inputLocator.elementHandle();
```
Comment (Janpot) [inline test/integration/propControls/index.spec.ts:22]: One potential way of avoiding relying on these implementation details is to add `Typography` elements in the same row and [layout](https://playwright.dev/docs/selectors#selecting-elements-based-on-layout) selectors to find the associated input, regardless of the label

File: test/utils/locators.ts [hunk 9]
```diff
@@ -1,3 +1,9 @@
 export function toolpadHomeAppRow(appName: string): string {
```
Comment (Janpot) [inline test/utils/locators.ts:1]: We should get rid of this file and move the locators inside of the `ToolpadHome` and `ToolpadEditor` models

File: docker/compose/docker-compose.test.yml (hunk 7 above)
Comment (apedroferreira) [reply docker/compose/docker-compose.test.yml:1]: yeah i didn't clean this up yet, just rebased

File: test/integration/propControls/index.spec.ts (hunk 8 above)
Comment (apedroferreira) [reply test/integration/propControls/index.spec.ts:22]: i can do that, i though using a label's `for` and the input `id` was actually better

File: test/utils/locators.ts (hunk 9 above)
Comment (apedroferreira) [reply test/utils/locators.ts:1]: yeah im gonna use those new objects, just havent done it yet

File: test/integration/propControls/index.spec.ts (hunk 8 above)
Comment (apedroferreira) [reply test/integration/propControls/index.spec.ts:17]: this is using accessibility props that should always match so i think it's a good approach (nothing implementation specific imo), we can change it if you insist though

Comment (apedroferreira) [reply test/integration/propControls/index.spec.ts:17]: but the label should always match the id, right? that's how html inputs should work.

Comment (apedroferreira) [reply test/integration/propControls/index.spec.ts:17]: or if it's an issue that we are kind of assuming the MUI inputs and labels have accessibility properly set up we can do it in another way - but it feels safe to assume that those accessibility props work.

Comment (Janpot) [reply test/integration/propControls/index.spec.ts:17]: Yeah, I was thinking about the edge-cases
* can an input have an `aria-label`?
* if the `id` changes over the lifetime of the page, this locator will stop working as it only considers the id of the first time it executed

but that may be too far fetched

Comment (apedroferreira) [reply test/integration/propControls/index.spec.ts:17]: i'll give locators another quick try, it's a more readable approach at least - for some reason i had some issues when i first tried due to the iframe, but now i understand things a bit better so maybe it's easy

Comment (Janpot) [reply test/integration/propControls/index.spec.ts:17]: Yes, I haven't fully internalized them yet. The way I understand them is that they are like `document.queryAll`, only it's sort of a live collection. So if the underlying DOM changes, the content of the locators change with it. I feel like I haven't fully harnassed their power yet, it seems to sort of require you to think differently about selectors.

Comment (apedroferreira) [reply test/integration/propControls/index.spec.ts:17]: layout selectors aren't working, maybe it's because the label intersects with the input, they don't seem very predictable at all... let's probably keep this approach for now, i don't think it should ever cause any issues.

**Comment by apedroferreira** [comment]
@Janpot 
CI passed 3 times in a row, maybe it's ok to merge this now?
i had to dispatch events manually in Firefox, couldn't find a better way.
If these tests start being flaky later though we can skip them temporarily and i can work on them a bit more - locally they always pass for me but CI seems a bit more unpredictable.

**Comment by Janpot** [comment]
Is there a ticket in the playwright repo that we can link to that describes the faulty behavior in firefox? If not, have you tried creating a minimal reproduction? Perhaps you can open a ticket and link to it?

I don't like to add hacks for bugs without adding documentation that can tell us more about the status of said bug. (example: https://github.com/mui/mui-toolpad/blob/4f71c00b28c8e57d9bad5ff99ede85f1d1aecc45/packages/toolpad-app/src/components/SplitPane.tsx#L58)

**Comment by apedroferreira** [comment]
> Is there a ticket in the playwright repo that we can link to that describes the faulty behavior in firefox? If not, have you tried creating a minimal reproduction? Perhaps you can open a ticket and link to it?
> 
> I don't like to add hacks for bugs without adding documentation that can tell us more about the status of said bug. (example:
> 
> https://github.com/mui/mui-toolpad/blob/4f71c00b28c8e57d9bad5ff99ede85f1d1aecc45/packages/toolpad-app/src/components/SplitPane.tsx#L58
> 
> )
> When there's a proper way forward towards solving the firefox issue in the future, then I'm fine merging this.

i couldn't find a proper related issue in the Playwright repo, and yeah i thought about create one myself, just takes a bit of work to create a reproduction. i'm working on it, there's probably no rush to merge this so we can wait

**Comment by apedroferreira** [comment]
> Is there a ticket in the playwright repo that we can link to that describes the faulty behavior in firefox? If not, have you tried creating a minimal reproduction? Perhaps you can open a ticket and link to it?
> 
> I don't like to add hacks for bugs without adding documentation that can tell us more about the status of said bug. (example:
> 
> https://github.com/mui/mui-toolpad/blob/4f71c00b28c8e57d9bad5ff99ede85f1d1aecc45/packages/toolpad-app/src/components/SplitPane.tsx#L58
> 
> )
> When there's a proper way forward towards solving the firefox issue in the future, then I'm fine merging this.

I've managed to reproduce the bug in a minimal reproduction:
https://github.com/microsoft/playwright/issues/17441

So it's confirmed, seems like there's a specific issue with triggering HTML5 drag & drop events in iFrames, in Firefox only.

I've included a link to the issue I created now, so will merge this PR if the tests seem stable in CI.

**Comment by Janpot** [comment]
great 👌

**Comment by Janpot** [comment]
@apedroferreira Following up on the playwright issue, it seems that it's closed due to the closing of https://github.com/microsoft/playwright/issues/17153 (would have been nice of them to reference that issue). Does calling `hover()` twice fix the issue?

**Comment by apedroferreira** [comment]
> @apedroferreira Following up on the playwright issue, it seems that it's closed due to the closing of [microsoft/playwright#17153](https://github.com/microsoft/playwright/issues/17153) (would have been nice of them to reference that issue). Does calling `hover()` twice fix the issue?

looks like they're supposed to have fixed it, i'll try removing the Firefox specific logic and see if things work now

**Comment by apedroferreira** [comment]
or i guess they might not have fixed it after all?
i'll try things from that issue then, maybe using `hover` twice works
